### PR TITLE
[PyTorch] Extend flatbuffer to support extra files map

### PR DESCRIPTION
**Pull request by pavithranrao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72951

Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

### CODE REVIEW

File: torch/csrc/jit/mobile/flatbuffer_loader.cpp [hunk 1]
```diff
@@ -172,6 +173,12 @@ mobile::Module FlatbufferLoader::parseModule(
   }
   IValue& module_ivalue = getIValue(module->state_obj());
 
+  auto extra_files_offsets = module_->extra_files();
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.cpp:176]: Let's pull this out as a separate function. i.e. parseModule should only parseModule; if you want extra files call a separater function that returns the file content.

File: torch/csrc/jit/mobile/flatbuffer_loader.h [hunk 2]
```diff
@@ -15,6 +15,7 @@
 namespace torch {
 namespace jit {
 
+using ExtraFilesMap = std::unordered_map<std::string, std::string>;
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.h:18]: if there an ExtraFilesMap definition somewhere else? If so reuse that one.

Comment (pavithranrao) [reply torch/csrc/jit/mobile/flatbuffer_loader.h:18]: It is not available while importing in mobile
 So it is redefined in [mobile/import.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/import.h#L14)
  
Once we integrate flatbuffer we can reuse the definition from mobile/import

Comment (qihqi) [reply torch/csrc/jit/mobile/flatbuffer_loader.h:18]: Sorry why cannot this file #include <torch/csrc/jit/mobile/import.h>? import.h is not part of mobile?

File: torch/csrc/jit/mobile/flatbuffer_loader.cpp [hunk 3]
```diff
@@ -172,6 +183,8 @@ mobile::Module FlatbufferLoader::parseModule(
   }
   IValue& module_ivalue = getIValue(module->state_obj());
 
+  parseExtraFiles(module, extra_files);
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.cpp:186]: Please expose this function as something users can call. and not call it here.

File: torch/csrc/jit/mobile/flatbuffer_loader.cpp [hunk 4]
```diff
@@ -543,22 +556,36 @@ TypePtr FlatbufferLoader::getOrCreateTypeAnnotations(
...
+      FlatbufferLoader().parseModule(flatbuffer_module, extra_files);
   m.set_delete_memory(std::move(data));
   return m;
 }
 
 mobile::Module initialize_mobile_module(
     mobile::serialization::Module* flatbuffer_module,
-    c10::optional<at::Device>) {
-  mobile::Module m = FlatbufferLoader().parseModule(flatbuffer_module);
+    ExtraFilesMap& extra_files,
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.cpp:570]: let's leave those function unchanged. Let user call a separate function for those who need extra files.

File: test/cpp/jit/test_flatbuffer.cpp [hunk 5]
```diff
@@ -137,6 +137,37 @@ TEST(FlatbufferTest, MethodInvocation) { // NOLINT (use =delete in gtest)
...
+
+  CompilationOptions options;
+  mobile::Module bc = jitModuleToMobile(*module, options);
+  auto buff = save_mobile_module_to_bytes(bc, extra_files);
+
+  std::unordered_map<std::string, std::string> loaded_extra_files;
+  loaded_extra_files["metadata.json"] = "";
+  auto* flatbuffer_module =
+      mobile::serialization::GetMutableModule(buff.data());
+  mobile::Module bc2 =
```
Comment (qihqi) [inline test/cpp/jit/test_flatbuffer.cpp:164]: this test no longer need to create a mobile::Module. We can call parseExtraFiles directly.

File: torch/csrc/jit/mobile/flatbuffer_loader.h [hunk 6]
```diff
@@ -39,13 +45,19 @@ TORCH_API mobile::Module initialize_mobile_module(
...
     c10::optional<at::Device> device = c10::nullopt);
 
 // Load a mobile::Module from a filepath.
 // This function does steps 1+2+3 described above.
 // We need to have this as a convienience because Python
 // API will need to wrap this. C++ clients should use one
 // versions above.
+TORCH_API mobile::Module load_mobile_module_from_file(
+    const std::string& filename,
+    ExtraFilesMap& extra_files,
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.h:58]: Remove this param. 

User who need extra file shall call parseExtraFiles separately.

**Review by iseeyuan — APPROVED** [review]
LGTM. Please address @qihqi 's comments before landing. 
To make the migration smooth, Please make sure that the at the top level APIs not changed, including the way to save/load extra files:

Python
Module._save_for_lite_interpreter(self, path), and _save_to_buffer_for_lite_interpreter  for saving and
_load_for_lite_interpreter(path_or_stream) for loading.
C++
jit::Module::_save_for_mobile(path) for saving and
_load_for_mobile(path_or_stream) for loading.

**Review by qihqi — APPROVED** [review]

**Pull request closed without merging**